### ISSUE TITLE
Enhanced depends logic: Convenience definition isn't entirely convenient

### ISSUE THREAD

**Issue by ezk84** [argoproj/argo-workflows#9370]
## Checklist

 

* [✅] Double-checked my configuration.
* [✅] Tested using the latest version.
* [✅] Used the Emissary executor.

## Summary

In [Enhanced Depends Logic](https://argoproj.github.io/argo-workflows/enhanced-depends-logic/) we see the following text:

> For convenience, if an omitted task result is equivalent to (task.Succeeded || task.Skipped || task.Daemoned).

This is confusing, but what it effectively means is that when I write `depends: A` that is equivalent to `depends: (A.Succeeded || A.Skipped || A.Daemoned)`. 

Recently, the Omitted status was made available to be part of these logical expressions. The trouble is now this creates the following scenarios:

1. If you skip a task A, and you reference that task in a dependent task B as `depends: A`, then B will run because having no extra status indicator implies `A.Succeeded || A.Skipped || A.Daemoned`. This is as intended.
2. If on the other hand a task B is omitted because an upstream dependency got skipped, and C depends on it as `depends: B` then C will *NOT* run because the above convenience definition doesn't include the Omitted status. 
 
Of course in this second case we can write C as depends: `B || B.Omitted` and then C will run as we hoped. This seems inconsistent as the difference between skipped and omitted is small and subtle.

How this looks: 

```
Case 1:
A [when: false] <- skipped
 \ 
  B [depends: A] <- runs

Case 2:
A [when: false] <- skipped
  \
   B [depends: A.Succeeded] <- omitted
   |
   C [depends: B] <- doesn't run unless we change to depends: B || B.Omitted
```
What version are you running?
3.4.0-rc1

## Diagnostics

Paste the smallest workflow that reproduces the bug. We must be able to run the workflow.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: inconsistent-depends-
  namespace: argo
spec:
  entrypoint: main
  templates:
    - name: main
      dag:
        tasks:
          - name: A
            when: 'false'
            template: pass
          - name: B
            depends: A.Succeeded
            template: pass
          - name: C
            depends: B
            template: pass
    - name: pass
      container:
        image: 'alpine:3.7'
        command:
          - sh
          - '-c'
          - exit 0
```

The expectation here is that task C runs also.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by tahiraha** [comment]
I'm still seeing this issue, but the Omitted state is not valid in enhanced depends logic in a DAG, has been documented here #10321

**Comment by agilgur5** [comment]
> In [Enhanced Depends Logic](https://argoproj.github.io/argo-workflows/enhanced-depends-logic/) we see the following text:
> 
> > For convenience, if an omitted task result is equivalent to (task.Succeeded || task.Skipped || task.Daemoned).
> 
> This is confusing, but what it effectively means is that when I write `depends: A` that is equivalent to `depends: (A.Succeeded || A.Skipped || A.Daemoned)`.

Yea that was _very_ confusing when using "omitted" to mean two different things; I fixed up the docs in #13110. PRs always welcome.

 

> Recently, the Omitted status was made available to be part of these logical expressions. The trouble is now this creates the following scenarios:

> Of course in this second case we can write C as depends: `B || B.Omitted` and then C will run as we hoped. This seems inconsistent as the difference between skipped and omitted is small and subtle.

I haven't used Workflows pre-`Omitted`, but I agree that it is a subtle difference. Unfortunately, changing that would break backward-compat. I think it's always best to explicitly write out all results anyway though.
